hammertime/rules/status.py

```diff
@@ -29,3 +39,123 @@ def __init__(self, *args):
     async def after_headers(self, entry):
         if entry.response.code in self.reject_set:
             raise RejectRequest("Status code reject: %s" % entry.response.code)
+
+
+class DetectSoft404:
+
+    def __init__(self):
+        self.engine = None
+        self.child_heuristics = Heuristics()
+        self.performed = defaultdict(dict)
+        self.soft_404_responses = defaultdict(list)
+
+    def set_engine(self, engine):
+        self.engine = engine
+
+    def set_kb(self, kb):
+        kb.soft_404_responses = self.soft_404_responses
+
+    async def after_response(self, entry):
+        server_address = urljoin(entry.request.url, "/")
+        request_url_pattern = self._extract_pattern_from_url(entry.request.url)
+        if server_address not in self.performed or request_url_pattern not in self.performed[server_address]:
+            # Temporarily assign a future to make sure work is not done twice
+            self.performed[server_address][request_url_pattern] = asyncio.Future()
+            response = await self._collect_sample(entry, request_url_pattern)
+            self.soft_404_responses[server_address].append(response)
+            self.performed[server_address][request_url_pattern].set_result(True)
+            # Remove the wait lock
+            self.performed[server_address][request_url_pattern] = None
+        elif self.performed[server_address][request_url_pattern] is not None:
+            await self.performed[server_address][request_url_pattern]
+
+        if entry.request.url == server_address:
+            return
+
+        for result in self.soft_404_responses[server_address]:
+            if result["pattern"] == request_url_pattern:
+                if result["code"] == entry.response.code and self._content_match(entry.response.content,
```

Comment:
Might be cleaner to push down the check for response code inside the matching method.